### PR TITLE
added cache support to form annotation builder

### PR DESCRIPTION
**Pull request by carnage**
As title. I'm sure there was an issue raised in relation to caching the output of the form builder; but I can't find it.

### CODE REVIEW

File: library/Zend/Form/Annotation/AnnotationBuilder.php [hunk 1]
```diff
@@ -83,6 +89,19 @@ class AnnotationBuilder implements EventManagerAwareInterface, FormFactoryAwareI
      */
     protected $preserveDefinedOrder = false;
 
+    public function setCache(StorageInterface $cache)
+    {
+        $this->cache = $cache;
+    }
+
+    /**
+     * @return \Zend\Cache\Storage\StorageInterface
```
Comment (malukenho) [inline library/Zend/Form/Annotation/AnnotationBuilder.php]:
```php
/**
 * @return StorageInterface
 */
```

File: library/Zend/Form/Annotation/AnnotationBuilder.php [hunk 2]
```diff
@@ -203,6 +222,17 @@ public function getFormSpecification($entity)
             }
         }
 
+        if (is_object($entity)) {
+            $cacheKey = 'FormAnnotationBuilder-' . get_class($entity);
+        } else {
+            $cacheKey = 'FormAnnotationBuilder-' . $entity;
+        }
```
Comment (malukenho) [inline library/Zend/Form/Annotation/AnnotationBuilder.php]: Maybe you can move it for another method, and you can test this one separated.

```php
$cacheKey = $this->generateKeyFor($entity)
```

Comment (carnage) [inline library/Zend/Form/Annotation/AnnotationBuilder.php]: I agree with this, I'll sort that out later

**Comment by weierophinney** [comment]
Caching of the forms created from annotations should be handled at the application layer, and not as a feature of the annotation builder:

```php
$form = $cache->getItem('YourFormName', $success);
if (! $success) {
    // build form using annotation builder
    $cache->setItem('YourFormName', $form);
}
```

The feature is not needed within the annotation builder itself.

**Comment by carnage** [comment]
I think you misunderstand what this is caching. It doesn't cache the forms themselves, but the intermediate form config array created by the expensive operation of parsing annoations. This representation is (should be) internal to the annotation builder. 

To access this to cache a form a user would have to call a different method on the builder cache that and then take care of creating the form using the factory themselves. Adding unnessisary steps just to get the same level of performance that building a form manually using array config would give them.

Caching intermediate forms of data structures is a pattern found embedded accross many frameworks which use annotations as meta data as the reading (using reflection) and parsing (regex's/string manipulation) is an expensive operation. Even PHP itself supports caching of its parsed form via APC - you wouldn't expect a user to take care of that in their applications would you?

Furthermore caching a form is something which is not going to work very well in all cases. Form elements, validators and filters are all fairly complex objects which could include uncacheable properties eg closures, db connections, request/user specific data.

I have been thinking about the structure of the parser in general as it is gaining too many responsibilties, however a refactor is not something that could be done for 2.4 and since 2.4 is supposed to be LTS, having caching support in is better than not having it at all because it polutes the architecture.

**Pull request closed without merging**